### PR TITLE
Initial draft for specificationa

### PR DESCRIPTION
**Pull request by OlegDokuka**
This PR provides an initial draft of the Reactive Streams Specification for JavaScript.

At the point, the following PR is aimed to amend the specification and after that provide work for actual implementation of Spec and TCK.

### CODE REVIEW

**Comment by OlegDokuka** [comment]
cc/ @viktorklang @kbahr @hypothete @rdegnan

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+## Copyright Statement
+
+The aforementioned waiver of copyrights and other rights is represented by the addition of a line to the file [CopyrightWaivers.txt](https://github.com/reactive-streams/reactive-streams-jvm/blob/master/CopyrightWaivers.txt). For a pull request to be considered every contributor must have signed the copyright statement in this way; this may be included within that same pull request.
+
+## Gatekeepers
+
+To ensure consistent development of Reactive Streams towards their goal, a group of gatekeepers is defined:
+
+* Kaazing Corp., currently represented by Todd Montgomery (@tmontgomery)
+* Netflix Inc., currently represented by Ben Christensen (@benjchristensen)
```
Comment (benlesh) [inline CONTRIBUTING.md:19]: @benjchristensen is at Facebook now. Perhaps you want @yunong?

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -0,0 +1,76 @@
+# Contributing to the Reactive Streams Project
```
Comment (benlesh) [inline CONTRIBUTING.md:1]: Add an appropriate code of conduct and link it prominently at the top of this file and in the README. Everyone should know that by interacting and contributing here they are agreeing to abide by the code of conduct. [Contributor Covenant](https://github.com/ContributorCovenant/contributor_covenant) is a widely used one. Not having a code of conduct will actually result in excluding people, which isn't what we want in open source.

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+The Reactive Streams project welcomes contributions from anybody who wants to participate in moving this initiative forward. All code or documentation that is contributed will have to be covered by a waiver of all copyrights and other rights as detailed by the LICENSE and COPYING files at each repository root, the rationale for this is that the APIs defined by this project shall be freely implementable and usable by everyone.
+
+## Copyright Statement
+
+The aforementioned waiver of copyrights and other rights is represented by the addition of a line to the file [CopyrightWaivers.txt](https://github.com/reactive-streams/reactive-streams-jvm/blob/master/CopyrightWaivers.txt). For a pull request to be considered every contributor must have signed the copyright statement in this way; this may be included within that same pull request.
```
Comment (benlesh) [inline CONTRIBUTING.md]: May want to move the copyright waivers to this repo so it does not get out of sync or moved.

File: tck/README.md [hunk 4]
```diff
@@ -0,0 +1 @@
+# Reactive Streams TCK #
```
Comment (benlesh) [inline tck/README.md]: - Add link to CONTRIBUTING doc
- Add link to a code of conduct.

**Review by benlesh — CHANGES_REQUESTED** [review]

File: tck/README.md (hunk 4 above)
Comment (OlegDokuka) [reply tck/README.md]: This just drafted a spec and not aimed to provide everything at once. Im planning to fix TCK readme once we agreed on spec. Then, after first step, going to fix docs. Does it sound good to you?

File: CONTRIBUTING.md (hunk 3 above)
Comment (OlegDokuka) [reply CONTRIBUTING.md]: Ah, right, I moved it, but missed that link. Thanks

File: CONTRIBUTING.md (hunk 2 above)
Comment (OlegDokuka) [reply CONTRIBUTING.md:1]: Mmm, makes sense. Followed RS spec JVM structure while writing everything, but anyway it makes sense 👍

File: CONTRIBUTING.md (hunk 1 above)
Comment (OlegDokuka) [reply CONTRIBUTING.md:19]: Good point. @viktorklang should we change it?

File: CONTRIBUTING.md (hunk 3 above)
Comment (OlegDokuka) [reply CONTRIBUTING.md]: fixed

File: CONTRIBUTING.md (hunk 2 above)
Comment (OlegDokuka) [reply CONTRIBUTING.md:1]: done

File: CONTRIBUTING.md (hunk 1 above)
Comment (kbahr) [reply CONTRIBUTING.md:19]: @yunong is no longer at Netflix either - I don't believe Netflix has *any* representation for this anymore. We should reach out to their engineering org to confirm. I'll follow up.

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+The Reactive Streams project welcomes contributions from anybody who wants to participate in moving this initiative forward. All code or documentation that is contributed will have to be covered by a waiver of all copyrights and other rights as detailed by the LICENSE and COPYING files at each repository root, the rationale for this is that the APIs defined by this project shall be freely implementable and usable by everyone.
+
+## Code of Conduct
+
+This project is governed by the [Code of Conduct](CODE_OF_CONDUCT.md).
+By participating you are expected to uphold this code.
+
+## Copyright Statement
+
+The aforementioned waiver of copyrights and other rights is represented by the addition of a line to the file [CopyrightWaivers.txt](https://github.com/reactive-streams/reactive-streams-jvm/blob/master/CopyrightWaivers.txt). For a pull request to be considered every contributor must have signed the copyright statement in this way; this may be included within that same pull request.
```
Comment (viktorklang) [inline CONTRIBUTING.md]: This needs to refer to the project-local CopyrightWaivers.txt

(currently refers to the JVM one) (https://github.com/reactive-streams/reactive-streams-jvm/blob/master/CopyrightWaivers.txt)

Comment (OlegDokuka) [reply CONTRIBUTING.md]: fixed

File: CONTRIBUTING.md (hunk 1 above)
Comment (viktorklang) [reply CONTRIBUTING.md:19]: I believe that most of that list is inaccurate right now, but let's leave it as is for now.

Comment (OlegDokuka) [reply CONTRIBUTING.md:19]: Sounds good

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+
+    add CONTRIBUTING.md
+
+    * clarify how pull requests should look like
+    * describe the acceptance process
+    * define the Copyright Statement signing process
+
+## Performing Official Releases
+
+Creating binary artifacts, uploading them to central repositories and declaring these to be an official release of the Reactive Streams project requires the consent of all gatekeepers. The process is initiated by creating a ticket in the `reactive-streams` repository for this purpose and consent is signaled in the same way as for pull requests. The actual work of updating version numbers and publishing the artifacts will typically involve pull requests targeting the affected repositories.
```
Comment (viktorklang) [inline CONTRIBUTING.md]: update `reactive-streams` to `reactive-streams-js`

File: README.md [hunk 7]
```diff
@@ -1,2 +1,300 @@
...
+| <a name="1.2">2</a>       | A `Publisher` MAY signal fewer `onNext` than requested and terminate the `Subscription` by calling `onComplete` or `onError`. |
+| [:bulb:](#1.2 "1.2 explained") | *The intent of this rule is to make it clear that a Publisher cannot guarantee that it will be able to produce the number of elements requested; it simply might not be able to produce them all; it may be in a failed state; it may be empty or otherwise already completed.* |
+| <a name="1.3">3</a>       | `onSubscribe`, `onNext`, `onError` and `onComplete` signaled to a `Subscriber` MUST be signaled in a [thread-safe](#term_thread-safe) manner—and if performed by multiple threads—use [external synchronization](#term_ext_sync). |
+| [:bulb:](#1.3 "1.3 explained") | *The intent of this rule is to make it clear that [external synchronization](#term_ext_sync) must be employed if the Publisher intends to send signals from multiple/different threads. Note, this rule is implemented by default in JavaScript, since it is [single-threaded](#term_single-threaded) within the same context* |
```
Comment (viktorklang) [inline README.md]: Missing period after `context`

File: CONTRIBUTING.md (hunk 6 above)
Comment (OlegDokuka) [reply CONTRIBUTING.md]: done

File: README.md [hunk 8]
```diff
@@ -1,2 +1,300 @@
...
+| <a name="2.6">6</a>       | A `Subscriber` MUST call `Subscription.cancel()` if the `Subscription` is no longer needed. |
+| [:bulb:](#2.6 "2.6 explained") | *The intent of this rule is to establish that Subscribers cannot just throw Subscriptions away when they are no longer needed, they have to call `cancel` so that resources held by that Subscription can be safely, and timely, reclaimed. An example of this would be a Subscriber which is only interested in a specific element, which would then cancel its Subscription to signal its completion to the Publisher.* |
+| <a name="2.7">7</a>       | A `Subscriber` MUST ensure that all calls on its `Subscription` take place from the same thread or provide for respective [external synchronization](#term_ext_sync). |
+| [:bulb:](#2.7 "2.7 explained") | *The intent of this rule is to establish that [external synchronization](#term_ext_sync) must be added if a Subscriber will be using a Subscription concurrently by two or more threads. Note, this rule is implemented by default in JavaScript, since it is [single-threaded](#term_single-threaded) within the same context* |
```
Comment (viktorklang) [inline README.md]: Missing period after context.

File: README.md [hunk 9]
```diff
@@ -1,2 +1,300 @@
...
+| <a name="2.10">10</a>     | A `Subscriber` MUST be prepared to receive an `onError` signal with or without a preceding `Subscription.request(n: number | bigint)` call. |
+| [:bulb:](#2.10 "2.10 explained") | *The intent of this rule is to establish that Publisher failures may be completely unrelated to signalled demand. This means that Subscribers do not need to poll to find out if the Publisher will not be able to fulfill its requests.* |
+| <a name="2.11">11</a>     | A `Subscriber` MUST make sure that all calls on its [signal](#term_signal) methods happen-before the processing of the respective signals. I.e. the Subscriber must take care of properly publishing the signal to its processing logic. |
+| [:bulb:](#2.11 "2.11 explained") | *The intent of this rule is to establish that it is the responsibility of the Subscriber implementation to make sure that asynchronous processing of its signals are thread safe. See [JMM definition of Happens-Before in section 17.4.5](https://docs.oracle.com/javase/specs/jls/se8/html/jls-17.html#jls-17.4.5). Note, this rule is implemented by default in JavaScript, since it is [single-threaded](#term_single-threaded) within the same context* |
```
Comment (viktorklang) [inline README.md]: Missing period after context.

Comment (OlegDokuka) [reply README.md]: fixed

File: README.md (hunk 8 above)
Comment (OlegDokuka) [reply README.md]: fixed

File: README.md (hunk 7 above)
Comment (OlegDokuka) [reply README.md]: fixed

**Comment by viktorklang** [comment]
@OlegDokuka It's a bit tricky to see the diff between the original README/spec and this one—mai I suggest adding the original first and creating the PR against it so the diff is "accurate"? :-)

**Comment by OlegDokuka** [comment]
@viktorklang sure thing. Give me a minute

**Comment by OlegDokuka** [comment]
@viktorklang done. Added additional commit as a copy/paste from reactive-streams-jvm, so now you can observe introduced changes in that way -> 
https://github.com/reactive-streams/reactive-streams-js/pull/5/files/d0b1e000b5baf7c5bbdf88b481fe020b716c841d..483069fc7f7cb1a9d2fa899545d4b29fa463a5ec

**Comment by viktorklang** [comment]
@OlegDokuka Perfect! Thank YOU! ^^

**Comment by viktorklang** [comment]
@reactive-streams/contributors Please have a look if you're interested!

**Comment by OlegDokuka** [comment]
@viktorklang dont have an access to that

**Comment by viktorklang** [comment]
@OlegDokuka you should be on @reactive-streams/js-contributors

**Comment by OlegDokuka** [comment]
# Long.MAX_VALUE in JS

Regarding usage of `BigInt` in JS. Unfortunately, JS supports only 2^53 as a MAX save the number to increment without losing calculation precision. Therefore, to keep spec binary compatible with other languages, we have to use `BigInt` type which allows doing math up to 10^1000. However, `BigInt` has its own performance penalty. What I measured is that increment of `BigInt` is about 5 times slower than the increment of the plain `Number`. For example, the following naive benchmark for `BigInt`:

```
function benchmark() {
    const startTime = process.hrtime();
    let i = 0n; // here is n notation after digit means BigInt type 

    while (i < 100000000) {
        i = i + 1n;
    }

    console.log(i);

    const endTime = process.hrtime();

    return [(endTime[0] - startTime[0]), endTime[1] - startTime[1]];
}
```

on average results in (on my 2.9 GHz Intel Core i9):

```
[ 4, 472734712 ] 
```

 > where the first number in the array is `Seconds` and the second in is `Nanos`.

In contrast, the same code but for a plain `Number` type on the same machine:

```
function benchmark() {
    const startTime = process.hrtime();
    let i = 0;

    while (i < 100000000) {
        i = i + 1;
    }

    console.log(i);

    const endTime = process.hrtime();

    return [(endTime[0] - startTime[0]), endTime[1] - startTime[1]];
}
```

on average results in:

```
[ 0, 47754006 ]
```
which is about 4~5 times slower.

### From our conversation with @viktorklang on Twitter

What I saw in Reactor Core JS, which was developed by @akarnokd: 

David Karnok, developing reactor-core-js implemented that special signal in JS as this -> https://github.com/reactor/reactor-core-js/blob/master/src/subscriber.js#L50

 
 Viktor Klang
 When is Infinity reached by adding numbers?

11m 10 minutes ago
 Oleh Dokuka

-> https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Infinity
 
 Viktor Klang
 I mean, if a Subscriber requests 1, and the Publisher doesn't publish anything (yet), when does the aggregation of demand reach the value Infinity?

6m 5 minutes ago
Guess more than iterating to Long.MAX_VALUE

the problem in js is that precision of calculation which overflows 2^53 becomes incorrect like the following

 Oleh Dokuka
 ```
> Number.MAX_SAFE_INTEGER + 1000013
9007199255741004
> Number.MAX_SAFE_INTEGER + 1000015
9007199255741006
> Number.MAX_SAFE_INTEGER + 1000016
9007199255741008
```

**Comment by OlegDokuka** [comment]
Ahh, even though we tried to rely on incorrect `Number` calculation, in case one send us `Long.MAX_VALUE - 1` we will not be able to do the right math. See the following sample -> https://jsbin.com/gujaxiqodu/edit?js,console

**Comment by rdegnan** [comment]
I would recommend keeping request_n as Number instead of BigInt. I don't see a need to preserve binary compatibility with other languages within a single language reactive streams implementation, that should be in the domain of binary reactive streams network protocols. RSocket for example limits request_n frames to 2^31-1, so it is up to the RSocket implementation to maintain high/low watermarks and hand out request_n frames up to the requested amount.

**Review by benlesh — APPROVED** [review]

**Comment by OlegDokuka** [comment]
After some thoughts, I concluded and agree with @rdegnan that we could rely on `Infinite` object as a signal for PUSH-only model and use `Number` instead of `BigInt` since it will have never been reached from the perspective of the time taken for incrementing to that `10^1000` value as well as from the JS behavior which prevents incrementing/decrementing for `Number` type after some specific number.

Any thoughts @viktorklang @benlesh?

**Comment by viktorklang** [comment]
@OlegDokuka Hmmm, if there is to be a bridge between implementations in different languages, think for instance in Graal, then that bridge would have to know to translate between Long.MAX_VALUE and Infinity. Do we htink that is going to be an issue?

**Comment by OlegDokuka** [comment]
@viktorklang I don't think this should be a big deal. I guess if such interop is somewhen, we can always provide an additional bridge for that which will catch `Infinite` from js and map it to Long.MAX_VALUE and vice versa.

**Comment by viktorklang** [comment]
@OlegDokuka The thing is that such a bridge would have to keep track of cumulative demand which overflows Long.MAX_VALUE too.

**Comment by OlegDokuka** [comment]
Another option - stick to `Long.MAX_VALUE` and leave details of the implementation for libs vendors. Anyway, using pure `Number,` it will not be possible to accumulate/decrement the requested amount by `one` started from `2^54`.

On the other hand, in real-world scenarios - no-one will request data by `one` up to `Long.MAX_VALUE`. If such happens - either publisher should wait for 300 years (or in case of JS for half of the year) which is less possible ( guess such publisher should be considered as dead and closed by timeout) or size of request should be higher than 1 or 10 or 100, so in such case, if client increment by 1000 and sending speed is 1000 lower than request frequency, then at some point sending model will be considered as pure push model and in that way interop will be achieved

**Comment by viktorklang** [comment]
@OlegDokuka What happens if someone asks for (Long.MAX_VALUE / 10) 10 times in a row?

**Comment by OlegDokuka** [comment]
https://jsbin.com/nariqatexe/edit?js,console

that is funny)

**Comment by OlegDokuka** [comment]
@viktorklang OK, what if we stick to long max value and leave details of the implementation for vendors. For example, in Reactor, we skip [validation]() of the negative request and provide some [StrictSubscriber](https://github.com/reactor/reactor-core/blob/master/reactor-core/src/main/java/reactor/core/publisher/StrictSubscriber.java) implementation which follows all the rules and applies to the given subscriber in case it is non-reactor ecosystem one

**Comment by OlegDokuka** [comment]
@viktorklang provided last fixes to the spec related to the upper limit. Spec keeps clear a requirement of the upper limit equal to `Long.MAX_VALUE`. It does not specify a particular type of solution and does not limit API to `bigint` type. It mentions JS limits and advice vendors to provide appropriate math in order to achieve the right precision in tracking the requested amount.

If it looks good to you, we can merge it and I will send TCK shortly

**Comment by viktorklang** [comment]
@OlegDokuka Latest commit looks good to me. You might want to check the latest master in reactive-streams-jvm for some updated verbiage, such as: https://github.com/reactive-streams/reactive-streams-jvm/commit/4d6cdb76babbd88c80303115b4d35c4343233da4

Nice work, Oleg!

**Comment by OlegDokuka** [comment]
Waits resolution for https://github.com/reactive-streams/reactive-streams-jvm/pull/448 so once we have it, amend latest fixes and merge this PR

**Pull request merged**